synapse/rest/client/v2_alpha/sync.py

```diff
@@ -229,25 +229,54 @@ async def encode_response(self, time_now, sync_result, access_token_id, filter):
         )
 
         logger.debug("building sync response dict")
-        return {
-            "account_data": {"events": sync_result.account_data},
-            "to_device": {"events": sync_result.to_device},
-            "device_lists": {
-                "changed": list(sync_result.device_lists.changed),
-                "left": list(sync_result.device_lists.left),
-            },
-            "presence": SyncRestServlet.encode_presence(sync_result.presence, time_now),
-            "rooms": {"join": joined, "invite": invited, "leave": archived},
-            "groups": {
-                "join": sync_result.groups.join,
-                "invite": sync_result.groups.invite,
-                "leave": sync_result.groups.leave,
-            },
-            "device_one_time_keys_count": sync_result.device_one_time_keys_count,
-            "org.matrix.msc2732.device_unused_fallback_key_types": sync_result.device_unused_fallback_key_types,
+        response = {
             "next_batch": await sync_result.next_batch.to_string(self.store),
         }
 
+        if sync_result.account_data:
+            response["account_data"] = {"events": sync_result.account_data}
+        if sync_result.presence:
+            response["presence"] = SyncRestServlet.encode_presence(
+                sync_result.presence, time_now
+            )
+
+        if sync_result.to_device:
+            response["to_device"] = {"events": sync_result.to_device}
+
+        if sync_result.device_lists.changed:
+            response.update(
+                {"device_lists": {"changed": list(sync_result.device_lists.changed)}}
+            )
+        if sync_result.device_lists.left:
+            response.update(
+                {"device_lists": {"left": list(sync_result.device_lists.left)}}
+            )
+
+        if sync_result.device_one_time_keys_count:
+            response[
+                "device_one_time_keys_count"
+            ] = sync_result.device_one_time_keys_count
+        if sync_result.device_unused_fallback_key_types:
+            response[
+                "org.matrix.msc2732.device_unused_fallback_key_types"
+            ] = sync_result.device_unused_fallback_key_types
+
+        if joined:
+            response.update({"rooms": {"join": joined}})
+        if invited:
+            response.update({"rooms": {"invite": invited}})
```

Comment:
Wouldn't this overwrite the existing `rooms` if you both joined a room and were invited to a room since last sync?

Reply:
I think you are right about that, I'll switch it to a defaultdict.

Reply:
Should be better now.